### DISCUSSION TITLE
Feature suggestion: round-robin collection type

### DISCUSSION THREAD

**Discussion by whmeitzler** [pocketbase/pocketbase#3398] [Q&A]
I am looking for a repeatable way to create collections that behave in a round-robin fashion:
- There is a maximum number of records in a collection
- The collection only allows insertion and read (not write or delete)
- In insert, if the collection is "full", the oldest record is overwritten by the new record

This is similar to [rrdtool](https://oss.oetiker.ch/rrdtool/)

Such a collection would be very useful to enable, for example, a graph of the last 1000 seconds of CPU usage. I think it meshes well enough with the realtime and single-daemon ethos of Pocektbase to warrant exploring making it a collection type. It could open up "time series" support without forcing a different storage solution than sqlite.

I'm exploring doing this with a master "rrd" table and creating collections that, if named `rrd_*` have hooks that exhibit these characteristics. It's a rather complex setup to achieve 2 simple goals:
  1. ensure a "sequence number" column I can use to find the correct row to insert
  2. provide a place to store the max size of the collection

I suspect that a new collection type could be the best way to do this, but I want to open a discussion before mucking too far in Pocketbase's internals. 
- The insert-and-read-only semantics mean such a collection type can use a numeric primary key, saving space and fulfilling the "sequence" requirement. 
- Size settings can be stored in the collection's "settings" JSON field
- The Admin UI can largely stay the same

**Comment by ganigeorgiev** [comment]
There no plans for now to introduce new collection types. I'm planning a small refactoring of the internals in the future to allow more "native" SQLite dev experience and I want to limit the exposed API surface for now.

Note that for the autoincrement/sequence number you can also use the builtin SQLite [rowid](https://www.sqlite.org/rowidtable.html) which is available with every collection table (_it is not exposed currently in the web APIs but it is accessible when using raw SQL in the hooks_).

**Comment by khrome83** [comment]
Curious why you describe this as round robin?

Are there multiple collections in your example @whmeitzler?

As far as limiting the number of rows, could you not just use the timestamp of when it was created, and the number of rows to identify if the oldest row needs to be removed first?

Ideally you could do this as a trigger in the database itself by using a migration (Go use Pocketbase as a framework).

When insert, check row count, if max, remove oldest row, then handle insertion. 

Not sure if it would work or not as a trigger, but that is where I would start so it's removed from application logic.

**Comment by ahamidi** [comment]
This is more of a [Capped Collection](https://www.mongodb.com/docs/manual/core/capped-collections/) and although I haven't tried myself, you can probably achieve it right now using rules + hooks.

**Comment by whmeitzler** [comment]
> There no plans for now to introduce new collection types. I'm planning a small refactoring of the internals in the future to allow more "native" SQLite dev experience and I want to limit the exposed API surface for now.


Fair enough. 

> Note that for the autoincrement/sequence number you can also use the builtin SQLite [rowid](https://www.sqlite.org/rowidtable.html) which is available with every collection table (it is not exposed currently in the web APIs but it is accessible when using raw SQL in the hooks).


That was a very helpful suggestion. Thanks :) 

> This is more of a [Capped Collection](https://www.mongodb.com/docs/manual/core/capped-collections/) and although I haven't tried myself, you can probably achieve it right now using rules + hooks.

You are correct, that's what I'm going for. 
I took a swing at implementing this with hooks:
https://github.com/whmeitzler/pocketbase-rrd/

Feedback is welcome, this is a good excuse to get better at using Pocketbase as a framework.
Thanks again @ganigeorgiev for sharing your sandbox and letting us play!
